app/scripts/models/account.js

```diff
@@ -298,24 +312,37 @@ define([
           return fxaClient.changePassword(email, oldPassword, newPassword);
         })
         .then(function () {
-          // sign the user in, keeping the current sessionTokenContext. This
-          // prevents sync users from seeing the `sign out` button on the
-          // settings screen.
           return fxaClient.signIn(
             email,
             newPassword,
             relier,
             {
-              reason: fxaClient.SIGNIN_REASON.PASSWORD_CHANGE,
-              sessionTokenContext: self.get('sessionTokenContext')
+              reason: fxaClient.SIGNIN_REASON.PASSWORD_CHANGE
             }
           );
         })
         .then(function (updatedSessionData) {
-          self.set(updatedSessionData);
+          self.storeSessionData(updatedSessionData, relier);
         });
     },
 
+    storeSessionData: function (sessionData, relier) {
+      var self = this;
+
+      sessionData.isSessionSharedWithFirefox =
+        self.get('isSessionSharedWithFirefox') || relier.isSync();
+
```

Comment:
I want to move [this branch from fxa-client.js](https://github.com/mozilla/fxa-content-server/pull/2853/files#diff-341971be75d35dbd724949cda527a0efR127) here too, but that bloated the PR more. I'll do some other cleanup in a follow on. Down with fxa-client.